### ISSUE TITLE
JSON didn't match expected structure. 'tag_name' missing

### ISSUE THREAD

**Issue by deviashin** [yknivag/ESP_OTA_GitHub_Showcase#2]
Please tell me why I see this?
JSON didn't match expected structure. 'tag_name' missing

**Comment by deviashin** [comment]
Having spent all day, I came up with just another error: Failed to parse JSON.

#define GHOTA_CURRENT_TAG "2.0.0"

https://github.com/deviashin/k_485_802_11_ota/releases/tag/2.2.0

What am I doing wrong?

**Comment by yknivag** [comment]
Hi,

Thanks for using the library.

Generally there are three reasons why these errors occur:

1. Trying to upgrade from a private repository (the library doesn't support authentication with the API).

2. Trying to upgrade from a repository with no releases or a latest release with no tag.

3. Going beyond the GitHub API usage allowance. Never put the upgrade code in your loop as GitHub only allow a small number of API requests per hour per IP. I ran into it several times testing this. Sometimes you just need to wait a while for GitHub to un-blacklist your IP.

Without seeing the full code it is difficult to know if there is something else you have done that has caused a problem.

The API response I get for your repository looks ok (https://api.github.com/repos/deviashin/k_485_802_11_ota/releases/latest). If you see an error message on that link then your IP is likely blocked and you will need to wait for it to be delisted.

Hope that helps.

**Comment by deviashin** [comment]
By the link https://api.github.com/repos/deviashin/k_485_802_11_ota/releases/latest I see a json file with a bunch of parameters. So the release was created correctly and the problem is not in it and not in blocking the ip address?

In your example, I get the same problem: https://github.com/yknivag/ESP_OTA_GitHub_Showcase/issues/1

Version 2.6.3 is installed in the board manager for esp8266

**Comment by yknivag** [comment]
You get the same error using this showcase example?

Which version of the ArduinoJSON library do you have installed?

**Comment by deviashin** [comment]
ArduinoJson by Benoit Blanchon v6.14.1

**Comment by deviashin** [comment]
I'm sorry, for some reason the link to the problem with your example was inserted incorrectly, there I get the Wrong HTTP Code
https://github.com/yknivag/ESP_OTA_GitHub_Showcase/issues/1

**Comment by yknivag** [comment]
Ok. Let me try and replicate with those exact versions and get back to you. Will be tomorrow now as I am busy this evening I'm afraid.

**Comment by deviashin** [comment]
Ok thanks so much for the help

**Comment by deviashin** [comment]
If it will not be difficult for you, could you please write the library versions with which everything exactly worked, I will try to check it myself and, possibly, save your precious time

**Comment by yknivag** [comment]
Hi,

I have been able to reproduce the "wrong HTTP code" error (GitHub had changed the format of their redirect headers).  I have a fix and will create a new version of the library this afternoon.

With the fixed version I am able to make this example retrieve the correct upgrade URL from your repository, so it is correctly retrieving and parsing the JSON.  I hope you'll understand that I will not upgrade my own board to your binary as I do not know what code you have in it.

I will update this issue again once I have uploaded the new version of the library.

Thanks,

Gavin.

For your reference, the versions I am using:

Arduino v.1.8.10
ArduinoJSON v.6.14.1
ESP82266 Boards v.2.6.3

**Comment by yknivag** [comment]
The "wrong HTTP code" issue has been fixed in release v 0.0.3 of the library which should be available to update in the Arduino library manager shortly.

I am not sure if this will fix the specific issue you were having but, as above, using the example code in the library I can get as far as performing the update using your repository.

Please let me know how you get on.

**Comment by deviashin** [comment]
Hello! Thanks for the corrections in the library, your sketch from the example, but using my personal repository (uploaded your binary file to my repository). At the first start, I upload the certificate to the file system, update the sketch and it successfully finds the new version, downloads and installs. Ok, that's cool!
However, after successful flashing OTA, the following information appears in the port monitor: Number of CA certs read: 0
It turns out that downloading the firmware overwrites SPIFFS?

**Comment by yknivag** [comment]
Ok so you're getting an update ok. That's good and it means that the library is now working as intended.

With regards to the issue with SPIFFS, this library uses the official [ESP8266httpUpdate library](https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266httpUpdate) to actually perform the update. It is important to note this because it is important to note that all the caveats to using that library also apply to this one.

The most notable caveat being that in order to preserve SPIFFS during update the OTA binary must be compiled with the same SRAM partition scheme (and the same board definition) as the sketch you are updating from. Otherwise it is possible for the code to write over part of the SPIFFS partition or may indicate to the uC that the SPIFFS partition starts and ends in a place it doesn't. This will stop the FS.h library reading SPIFFS.

tl;dr make sure you use the same board and memory settings for the new sketch as the one you're updating from was compiled with.

**Comment by deviashin** [comment]
You were right, this is really stupid carelessness. 
I checked all the settings, made a new release in the repository and everything worked out!
The problem is completely resolved!
Thank you for your help and a good library!

**Comment by yknivag** [comment]
Getting the memory map wrong is a super easy thing to do. I can't tell you how many times I did it during development.

I'm glad you have it working and I hope it proves useful to you.
